Introduce New Hooks & Methods in Tumblr Importer

emdashcodes opened the pull request:
This PR introduces new hooks and methods, making the importer more flexible and extensible. The purpose of these additions is to simplify customization for developers, allowing them to handle different use cases more efficiently without altering the core importer.

## Testing Instructions:
* Run this branch and test the connection from scratch:
     * Clear any saved data from the plugin.
     * Follow the prompts to reconnect the OAuth App.
* Verify functionality:
    * Ensure all screens display as expected.
    * Check that imports function correctly.
    * Test importing blogs with various types of content and configurations.

Commit:
Make the importer more extendable

Commit:
Additional linter fixes

Commit:
comment spacing

Commit:
Fix order of sanitization

zaerl commented on class-tumblr-import.php:
```diff
@@ -364,10 +412,52 @@ public function start_blog_import() {
 
 			$this->schedule_import_job( 'do_blog_import', array( $url ) );
+
+			do_action( 'import_start', 'Tumblr' );
```
Let's also add an `import_end` like in the [WordPress importer](https://github.com/WordPress/wordpress-importer/blob/master/src/class-wp-import.php#L154). What do you think?

emdashcodes replied:
Good call, done!

zaerl commented on class-tumblr-import.php:
```diff
@@ -639,16 +774,61 @@ public function do_pages_import( $url ) {
 		 * @param string $filename The filename.
 		 *
-		 * @return int|WP_Error
+		 * @return void|WP_Error
```
Shouldn't this be `int|WP_Error`? `media_handle_upload` [returns an ID](https://developer.wordpress.org/reference/functions/media_handle_sideload/#return).

emdashcodes replied:
Yeah good catch, fixed this!

rtio commented on class-tumblr-import.php:
```diff
@@ -34,8 +59,10 @@ public function start() {
-			@$this->secretkey   = defined( 'TUMBLR_SECRET_KEY' ) ? TUMBLR_SECRET_KEY : ( ! empty( $_POST['secretkey'] ) ? $_POST['secretkey'] : $this->secretkey );
+			do_action( 'tumblr_importer_import_start' );
+
+			@$this->consumerkey = defined( 'TUMBLR_CONSUMER_KEY' ) ? TUMBLR_CONSUMER_KEY : ( ! empty( $_POST['consumerkey'] ) ? sanitize_text_field( wp_unslash( $_POST['consumerkey'] ) ) : $this->consumerkey );
```
wondering how much trouble would be to avoid using `@`  to suppress errors? also side question, what is the minimum PHP versions supported by this plugin? I didn't find it easily

zaerl replied:
>  what is the minimum PHP versions supported by this plugin? I didn't find it easily

This importer requires at least WordPress 3.2, so it must be compatible with PHP 5.6.

emdashcodes replied:
I've removed the suppression here for now and I'm not seeing any issues with the line.

rtio commented on class-tumblr-import.php:
```diff
@@ -639,16 +774,61 @@ public function do_pages_import( $url ) {
+		 *
+		 * @param  array $post the (already) imported WP_Post
+		 *
+		 * @return void
```
will `$this->{$sideload_method}( $post );`  never return? if those methods don't return anything you could just call it without `return`

emdashcodes replied:
Good catch. None of these methods return anything, they just make updates.

rtio reviewed (commented):
Just tested and it works as expected.

tommusrhodus commented on class-tumblr-import.php:
```diff
@@ -378,7 +468,7 @@ public function start_blog_import() {
 			check_admin_referer( 'tumblr-import' );
 			delete_option( get_class( $this ) );
-			wp_redirect( '?import=tumblr' );
+			wp_safe_redirect( '?import=tumblr' );
```
This should be followed by `exit;`

emdashcodes replied:
Fixed!

tommusrhodus commented on class-tumblr-import.php:
```diff
@@ -482,8 +604,15 @@ public function do_posts_import( $url ) {
+					do_action( 'tumblr_importer_importing_post_after', $post );
+
+					do_action( 'tumblr_importer_insert_new_post_before', $post );
 					$id = wp_insert_post( $post );
```
Thoughts on turning the third parameter of this function `false`?

https://developer.wordpress.org/reference/functions/wp_insert_post/

Might help avoid things like social plugins firing off 500 updates as posts are inserted.

emdashcodes replied:
I think that's a good idea. Updated both places.

tommusrhodus commented on class-tumblr-import.php:
```diff
@@ -639,16 +774,61 @@ public function do_pages_import( $url ) {
+			do_action( 'tumblr_importer_metadata_after', $post );
+
+			do_action( 'tumblr_importer_sideload_wp_update_post_before', $post );
+			wp_update_post( $post );
```
Same here, I think turning off hooks would be a good idea

https://developer.wordpress.org/reference/functions/wp_update_post/

Commit:
Handle PR Feedback

emdashcodes commented:
@zaerl @rtio @tommusrhodus Thank you all for the review! I've pushed some changes addressing your feedback and have tested against .org and .com again 👍

emdashcodes commented:
@rtio @zaerl Does this look okay to ya'll to merge? ❤️

rtio reviewed (approved):
LGTM :shipit:

zaerl commented:
It's ok to me.
